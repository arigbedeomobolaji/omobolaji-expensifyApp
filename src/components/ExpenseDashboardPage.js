//jshint ignore:start

import React from "react"
import ConnectExpenseList from "./ExpenseList"
import ConnectedExpenseListFilter from "./ExpenseListFilter"
import ConnectedExpenseSummary from "./ExpenseSummary"

const ExpenseDashboardPage = () => (
 <div>
  <ConnectedExpenseSummary />
  <ConnectedExpenseListFilter />
  <ConnectExpenseList />
 </div>
)

export default ExpenseDashboardPage