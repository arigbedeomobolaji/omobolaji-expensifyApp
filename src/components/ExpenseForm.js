//jshint ignore:start
import React from "react"
import moment from "moment"
import 'react-dates/initialize'
import { SingleDatePicker } from "react-dates"


export default class ExpenseForm extends React.Component {
 constructor(props){
  super(props)
  this.state = {
   description: props.expense ? props.expense.description : "",
   note: props.expense ? props.expense.note :"",
   amount: props.expense ? (props.expense.amount/100).toString() : "",
   createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
   isCalendarFocused: false,
   error: ""
  }
 }

 onDescriptionChange = (e) => {
  const description = e.target.value
  this.setState(() => ({
   description
  }))
 }

 onNoteChange = (e) => {
  const note = e.target.value
  this.setState(() => ({
   note
  }))
 }

 onAmountChange = (e) => {
  const amount = e.target.value
  if (amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
   this.setState(() => ({
    amount: amount
   }))
  }
 }

 onDateChange = (createdAt) => {
  if (createdAt) {
   this.setState(() => ({ createdAt }))
  }
 }
 

 onCalendarFocusedChange = ({ focused }) => {
  this.setState(() => ({
   isCalendarFocused: focused
  }))
 }

 onSubmit = (e) => {
  e.preventDefault()
  
  if (!this.state.description || !this.state.amount) {
   // Render Error 
   this.setState(() => ({
    error: "Please provide description and amount."
   }))
  } else {
   // clear error
   this.setState(() => ({
    error: ""
   }))
   // We created the expense props simply because we want to reuse this form on AddExpense and EditExpense
   // We can also call a function as a props
   this.props.onSubmit({
    description: this.state.description,
    amount: parseFloat(this.state.amount, 10) * 100,
    createdAt: this.state.createdAt.valueOf(),
    note: this.state.note
   })
  }
 }

 render() {
  return (
    <form className="form" onSubmit={this.onSubmit}>
     {!!this.state.error && <p className="form__error">{this.state.error}</p>}
     <input
      type="text"
      className="text-input"
      placeholder="Description"
      autoFocus
      value={this.state.description}
      onChange={this.onDescriptionChange}
     />
     <input 
      type="text"
      className="text-input"
      placeholder="Amount"
      value={this.state.amount}
      onChange={this.onAmountChange}
     />
     <SingleDatePicker 
      date={this.state.createdAt}
      onDateChange={this.onDateChange}
      focused={this.state.isCalendarFocused}
      onFocusChange={this.onCalendarFocusedChange}
      numberOfMonths={1}
      isOutsideRange={() => false}
     />
     <textarea
      className="textarea"
      placeholder="Add a Note (Optional)"
      value={this.state.note}
      onChange={this.onNoteChange}
     >
     </textarea>
     <div>
       <button className="button button--secondary" type="submit"> { !!this.props.buttonName ? this.props.buttonName : "Add Expense"}</button>
     </div>
    </form>
  )
 }
}