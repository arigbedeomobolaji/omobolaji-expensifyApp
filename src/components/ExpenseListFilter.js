//jshint ignore: start
import React from "react"
import { connect } from "react-redux"
import "react-dates/initialize"
import { DateRangePicker } from "react-dates"
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate } from "../actions/filters.action"
import "react-dates/lib/css/_datepicker.css"

export class ExpenseListFilter extends React.Component {

  state = {
  calendarFocused: null
 }

 onDatesChange = ({ startDate, endDate }) => {
  this.props.setStartDate(startDate)
  this.props.setEndDate(endDate)
  }
  
  onFocusChange = (calendarFocused) => this.setState(() => (
    { calendarFocused }
  ))
    
  onTextFilterChange = (e) => {
    this.props.setTextFilter(e.target.value)
  }
    
  onSortByChange = (e) => {
    e.target.value === "date"? this.props.sortByDate() : this.props.sortByAmount()
  }

  render() {  
    return (
    <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
            type="text"
            className="text-input"
            placeholder="Search expenses"
            //Any time you see value like below they are called CONTROLLED INPUT ==> Input controlled by Javascript
            value={this.props.filters.text}
            onChange={this.onTextFilterChange} 
            />
          </div>
          <div className="input-group__item">
            <select
            className="select"
            onChange={this.onSortByChange}
            >
            <option value="date">Date</option>
            <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker 
            startDate={this.props.filters.startDate}
            startDateId={"myStartDateId"}
            endDate={this.props.filters.endDate}
            endDateId={"myEndDateId"}
            onDatesChange={this.onDatesChange}
            focusedInput={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            isOutsideRange={() => false}
            showClearDates={true}
            />
          </div>
      </div>
    </div>
    )
 }
}

const mapStateToProps = (state) => {
 return {
  filters: state.filters
 }
}

const mapDispatchToProps = (dispatch) => ({
  setStartDate: startDate => dispatch(setStartDate(startDate)),
  setEndDate: endDate => dispatch(setEndDate(endDate)),
  setTextFilter: text => dispatch(setTextFilter(text)),
  sortByDate: () => dispatch(sortByDate()),
  sortByAmount: () => dispatch(sortByAmount())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilter)