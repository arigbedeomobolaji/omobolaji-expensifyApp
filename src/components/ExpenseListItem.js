//jshint ignore:start
import React from "react"
import { Link } from "react-router-dom"
import { connect } from "react-redux"

export const ExpenseListItem = ({dispatch, id, description, amount, createdAt}) => (
 <tr>
    <td>
      <Link to={`/edit/${id}`}>{description}</Link>
    </td>
    <td>{amount}</td>
    <td>{createdAt}</td>
 </tr>
)

export default connect()(ExpenseListItem) 