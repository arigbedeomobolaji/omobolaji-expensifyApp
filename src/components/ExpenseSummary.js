//jshint ignore:start
import React from "react"
import { connect } from "react-redux"
import {Link} from "react-router-dom"
import numeral from "numeral"
import selectExpenseTotal from "../selectors/expense-total"
import selectExpense from "../selectors/expenses.selector"

export const ExpenseSummary = ({expenseCount, expensesTotal}) => {
 const expenseWord = expenseCount === 1 ? "expense" : "expenses"
 const expenseTotalCurrency = numeral(expensesTotal / 100).format("$0,0.00")
 
  return (
    <div className="page-header">
      <div className="content-container">
        {
        expenseCount > 0 && <h1 className="page-header__title"> Viewing <span>{expenseCount}</span> {expenseWord} totalling <span>{expenseTotalCurrency}</span> </h1> 
        }
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Expense</Link>
        </div>
      </div>
 </div>
)}

const mapStateToProps = (state) => ({
 expenseCount: selectExpense(state.expenses, state.filters).length,
 expensesTotal: selectExpenseTotal(selectExpense(state.expenses, state.filters))
})


export default connect(mapStateToProps)(ExpenseSummary)