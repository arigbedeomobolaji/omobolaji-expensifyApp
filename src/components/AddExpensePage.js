//jshint ignore:start
import React from "react"
import {connect} from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {asyncAddExpense} from "../actions/expenses.action.js"

// History is passed into AddExpensePage becus AddExpensagePage is registered to a Route
export class AddExpensePage extends React.Component{
 onSubmit = (expense) => {
  this.props.asyncAddExpense(expense)
  this.props.history.push("/dashboard")
 }
 render() {
  return (
   <div>
    <div className="page-header">
     <div className="content-container">
      <h1 className="page-header__title">Add Expense</h1>
     </div>
    </div>
    <div className="content-container">
     <ExpenseForm
      onSubmit={this.onSubmit}
     />
    </div>
   </div>
  )
 }
}

const mapDispatchToProps = (dispatch) => ({
 asyncAddExpense: (expense) => dispatch(asyncAddExpense(expense))
})

export default connect(undefined, mapDispatchToProps)(AddExpensePage)

// mapDispatchToProps is a way to return your dispatcher functions allowing you to abstract them away from the component itself.
// to avoid inline functions we are going to switch our stateless function to a class based function