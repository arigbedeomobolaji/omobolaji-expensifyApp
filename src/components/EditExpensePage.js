//jshint ignore:start
import React from "react"
import { connect } from "react-redux"
import ExpenseForm from "./ExpenseForm"
import {asyncEditExpense, asyncRemoveExpense} from "../actions/expenses.action"

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    // props.dispatch(editExpense(props.expense.id, expense))
    this.props.asyncEditExpense(this.props.expense.id, expense)
    this.props.history.push("/dashboard")
  }

  handleRemoveExpense = (e) => {
    this.props.asyncRemoveExpense(this.props.expense.id)
    this.props.history.push("/dashboard")
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm
            expense={this.props.expense}
            onSubmit={this.onSubmit}
            buttonName={"Save Expense"}
          />
          <button className="button button-secondary" onClick={this.handleRemoveExpense}
          >Remove</button>
      </div>
      </div>
    )
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find((expense) => expense.id === props.match.params.id)   
})

const mapDispatchToProps = (dispatch) => ({
  asyncEditExpense: (id, expense) => {
    dispatch(asyncEditExpense(id, expense))
  },
  asyncRemoveExpense: (id) => {
    dispatch(asyncRemoveExpense(id))
  }
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage)
