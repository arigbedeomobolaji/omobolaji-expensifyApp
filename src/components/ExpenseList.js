//jshint ignore:start

import React from "react"
import { connect } from "react-redux"
import ExpenseListItem from "./ExpenseListItem"
import selectExpenses from "../selectors/expenses.selector"

export const ExpenseList = (props) => (
 <div>
  <h1>ExpenseList</h1>
  {
   props.expenses.length === 0 ? ( <p>No Expense to show</p>) : (
    
  <table>
   <thead>
    <tr>
     <th>Description</th>
     <th>Amount</th>
     <th>createdAt</th>
    </tr>
   </thead>
   <tbody>
    {props.expenses.map(
     (expense) => {
      return <ExpenseListItem key={expense.id} {...expense} />
     }
    )}
   </tbody>
  </table>
   )
  }
 </div>
)

const mapStateToProps = (state) => ({
 expenses: selectExpenses(state.expenses, state.filters)
})

export default connect(mapStateToProps)(ExpenseList)