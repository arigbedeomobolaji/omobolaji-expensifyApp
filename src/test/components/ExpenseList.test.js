// When we are testing react component we want to make sure we are testing the unconnected version because we want to be able to provide a set of dynamic props.
//jshint ignore: start
import React from "react"
import { shallow } from "enzyme"
import { ExpenseList } from "../../components/ExpenseList"
import expenses from "../fixtures/expense" 

test("should render ExpenseList with expenses correctly", () => {
 const wrapper = shallow(<ExpenseList expenses={expenses} />)
 expect(wrapper).toMatchSnapshot()
})

test("should render ExpenseList correctly with empty expenseList", () => {
 const wrapper = shallow(<ExpenseList expenses={[]} />)
 expect(wrapper).toMatchSnapshot()
})