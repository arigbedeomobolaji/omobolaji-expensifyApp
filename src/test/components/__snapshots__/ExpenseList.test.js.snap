// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseList correctly with empty expenseList 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Expense to show
      </span>
    </div>
  </div>
</div>
`;

exports[`should render ExpenseList with expenses correctly 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Expenses
    </div>
    <div
      className="show-for-desktop"
    >
      Expense
    </div>
    <div
      className="show-for-desktop"
    >
      Amount
    </div>
  </div>
  <div
    className="list-body"
  >
    <div>
      <Connect(ExpenseListItem)
        amount={5000}
        createdAt={0}
        description="Rent"
        id="1"
        key="1"
        note="My Rent for March"
      />
      <Connect(ExpenseListItem)
        amount={100}
        createdAt={-345600000}
        description="Gum"
        id="2"
        key="2"
        note="My chewing Gum for reading"
      />
      <Connect(ExpenseListItem)
        amount={500}
        createdAt={345600000}
        description="Coffee"
        id="3"
        key="3"
        note="My early morning coffee"
      />
    </div>
  </div>
</div>
`;
